### PR TITLE
Distillation loss implementation

### PR DESCRIPTION
**Pull request by shrsinh**
Distillation loss implementation along with the dataloader expecting input_ids, labels and teacher_logits

### CODE REVIEW

File: megatron/core/fusions/cce_loss.py [hunk 1]
```diff
@@ -36,6 +36,8 @@ def _maybe_build_vp_opts(vocab_size: int) -> Optional["VocabParallelOptions"]:
     group = get_tensor_model_parallel_group()
     return VocabParallelOptions.from_vocab(vocab_size, group=group)
     # return VocabParallelOptions(start, end, group=group)
+    return VocabParallelOptions.from_vocab(vocab_size, group=group)
+    # return VocabParallelOptions(start, end, group=group)
```
Comment (himanshu-livup) [inline megatron/core/fusions/cce_loss.py]: What's the change here? Revert any trivial changes involving only whitespaces.

File: megatron/core/fusions/cce_loss.py [hunk 2]
```diff
@@ -48,6 +50,8 @@ def cce_per_token_loss(
     reduction: str = "none",
     shift: bool = True,
     ignore_index: int = -100,
+    return_lse: bool = False,
+    temp: float = 1.0,
```
Comment (himanshu-livup) [inline megatron/core/fusions/cce_loss.py]: change variable name to temperature. temp is confusing and is generally used as short for temporary.

File: megatron/core/models/gpt/gpt_model.py [hunk 3]
```diff
@@ -367,9 +391,7 @@ def forward(
             runtime_gather_output (bool): Gather output at runtime. Default None means
                 `parallel_output` arg in the constructor will be used.
         """
-
```
Comment (himanshu-livup) [inline megatron/core/models/gpt/gpt_model.py:370]: Remove this whitespace change

File: megatron/core/models/gpt/gpt_model.py [hunk 4]
```diff
@@ -367,9 +391,7 @@ def forward(
...
         inference_context = deprecate_inference_params(inference_context, inference_params)
-
```
Comment (himanshu-livup) [inline megatron/core/models/gpt/gpt_model.py:372]: Remove this whitespace change

File: megatron/core/models/gpt/gpt_model.py [hunk 5]
```diff
@@ -473,6 +497,30 @@ def _postprocess(
...
+                        labels
+                    )
+                    traditional_kl_loss = traditional_distillation_loss(
+                        logits,
+                        teacher_data,
+                        labels_cp,
+                        T=self.config.distillation_temp,
+                        ignore_index=self.config.linear_ce_ignore_index,
+                    )
+
```
Comment (himanshu-livup) [inline megatron/core/models/gpt/gpt_model.py]: Indentation issue in this block. Use standard number of spaces for indentation similar to the code above.

File: megatron/core/models/gpt/gpt_model.py [hunk 6]
```diff
@@ -548,8 +596,8 @@ def _postprocess(
                 world_rank = torch.distributed.get_rank()
                 print(
                     f"[CCE debug] rank={world_rank} tp_rank={tp_rank} vocab_size={vocab_size}"
-                )
-
+                ) 
+
```
Comment (himanshu-livup) [inline megatron/core/models/gpt/gpt_model.py]: Remove this whitespace change.

File: megatron/core/transformer/transformer_config.py [hunk 7]
```diff
@@ -691,6 +691,23 @@ class TransformerConfig(ModelParallelConfig):
...
+    debug_distillation: bool = False
+
+    # For distillation loss
+    distillation_loss: bool = False
+
+    # Calculate using traditional method
+    distillation_with_traditional: bool = False
+
+    # Distillation Temperature
+    distillation_temp: float = 3.0
```
Comment (himanshu-livup) [inline megatron/core/transformer/transformer_config.py]: rename to distillation_temperature

File: megatron/core/utils.py [hunk 8]
```diff
@@ -1808,10 +1808,12 @@ def get_batch_on_this_cp_rank(batch: Dict[str, Any]):
     # we split sequence into 2*CP ranks. Assuming CP=2, we then get 4 chunks, chunk_0
     # and chunk_3 are assigned to GPU0, chunk_1 and chunk_2 are assigned to GPU1, so
     # that we can get balanced workload among GPUs in a context parallel group.
-    cp_size = parallel_state.get_context_parallel_world_size()
+    cp_size = parallel_state.get_context_parallel_world_size()
```
Comment (himanshu-livup) [inline megatron/core/utils.py]: remove whitespace change

File: megatron/training/arguments.py [hunk 9]
```diff
@@ -3169,3 +3170,19 @@ def _add_sft_args(parser):
...
                        help='SFT prompt format.')
     return parser
+
+
+def _add_distillation_args(parser):
+    group = parser.add_argument_group(title='distillation')
+    group.add_argument('--distillation-loss', action='store_true',
+                       help='Enable knowledge distillation during training. When enabled, '
+                       'data paths should reference JSONL directories containing teacher logits.')
+    group.add_argument('--distillation-temp', type=float, default=3.0,
```
Comment (himanshu-livup) [inline megatron/training/arguments.py]: rename to distillation-temperature

File: megatron/training/utils.py [hunk 10]
```diff
@@ -546,30 +572,30 @@ def _broadcast(item):
         tokens = torch.empty(
             (args.micro_batch_size, args.seq_length),
             dtype=torch.int64,
-            device=torch.cuda.current_device(),
+            device=device,
```
Comment (himanshu-livup) [inline megatron/training/utils.py]: No need for these changes. Let's avoid unnecessary changes in Megatron-LM codebase.

File: megatron/training/utils.py [hunk 11]
```diff
@@ -531,9 +560,6 @@ def _broadcast(item):
             _broadcast(batch['position_ids'])
 
         elif mpu.is_pipeline_last_stage():
-            # Multi-Token Prediction (MTP) layers need tokens and position_ids to calculate embedding.
-            # Currently the Multi-Token Prediction (MTP) layers is fixed on the last stage, so we need
-            # to broadcast tokens and position_ids to all of the tensor parallel ranks on the last stage.
```
Comment (himanshu-livup) [inline megatron/training/utils.py:536]: Don't remove these

File: megatron/training/utils.py [hunk 12]
```diff
@@ -546,30 +572,30 @@ def _broadcast(item):
...
             (args.micro_batch_size, args.seq_length),
             dtype=torch.int64,
-            device=torch.cuda.current_device(),
+            device=device,
         )
         labels = torch.empty(
             (args.micro_batch_size, args.seq_length),
             dtype=torch.int64,
-            device=torch.cuda.current_device(),
+            device=device,
```
Comment (himanshu-livup) [inline megatron/training/utils.py]: same as above. Do for all 5 occurences.

File: megatron/training/utils.py [hunk 13]
```diff
@@ -588,9 +614,6 @@ def _broadcast(item):
             _broadcast(position_ids)
 
         elif mpu.is_pipeline_last_stage():
-            # Multi-Token Prediction (MTP) layers need tokens and position_ids to calculate embedding.
-            # Currently the Multi-Token Prediction (MTP) layers is fixed on the last stage, so we need
-            # to broadcast tokens and position_ids to all of the tensor parallel ranks on the last stage.
```
Comment (himanshu-livup) [inline megatron/training/utils.py:593]: Don't remove this

File: megatron/training/utils.py [hunk 14]
```diff
@@ -499,7 +504,12 @@ def _broadcast(item):
                 group=mpu.get_tensor_model_parallel_group(),
             )
 
-    if mpu.get_tensor_model_parallel_rank() == 0:
+    is_tp0 = mpu.get_tensor_model_parallel_rank() == 0
+
+    teacher_header = torch.empty(2, dtype=torch.int64, device=device)
```
Comment (himanshu-livup) [inline megatron/training/utils.py]: Add all the teacher related code in this function under the distillation_loss flag.

File: megatron/training/utils.py [hunk 15]
```diff
@@ -610,6 +633,31 @@ def _broadcast(item):
...
+        _broadcast(teacher_tensors['values'])
+    else:
+        teacher_tensors = None
+
+    if args.pipeline_model_parallel_size != 1:
+        if mpu.is_pipeline_first_stage() or not mpu.is_pipeline_last_stage():
+            teacher_tensors = None
+
+    batch['teacher_data'] = teacher_tensors
+
```
Comment (himanshu-livup) [inline megatron/training/utils.py]: Add under distillation_loss flag

File: megatron/training/utils.py [hunk 16]
```diff
@@ -517,7 +527,26 @@ def _broadcast(item):
...
+                10,
+                device,
+            )
+        packed, shape = pack_teacher_batch(teacher_raw, args.seq_length, device)
+        if shape[0] < 0 or shape[1] < 0:
+            teacher_header.fill_(-1)
+        else:
+            teacher_header[0] = shape[0]
+            teacher_header[1] = shape[1]
+        teacher_tensors = packed
```
Comment (himanshu-livup) [inline megatron/training/utils.py]: Add under distillation_loss flag

File: pretrain_gpt.py [hunk 17]
```diff
@@ -43,16 +48,27 @@ def get_batch(data_iterator):
...
+        return None, None, None, None, None, None
 
     # get batches based on the TP rank you are on
     with mem_phase("LOAD_BATCH", do_barrier=True), nvtx_range("LOAD_BATCH"):
         batch = get_batch_on_this_tp_rank(data_iterator)
 
-        # slice batch along sequence dimension for context parallelism
-        batch = get_batch_on_this_cp_rank(batch)
+    # slice batch along sequence dimension for context parallelism
+    batch = get_batch_on_this_cp_rank(batch)
```
Comment (himanshu-livup) [inline pretrain_gpt.py]: Move it inside the with mem_phase block.

File: pretrain_gpt.py [hunk 18]
```diff
@@ -43,16 +48,27 @@ def get_batch(data_iterator):
...
+
+    teacher_packed = batch.pop('teacher_data', None)
+    teacher_data = None
+    if teacher_packed is not None and has_teacher_data(teacher_packed):
+        teacher_data = unpack_teacher_batch(teacher_packed)
```
Comment (himanshu-livup) [inline pretrain_gpt.py]:
```
teacher_data = None
if args.distillation_loss:
    teacher_packed = batch.pop('teacher_data', `None)`
    if teacher_packed is not None and has_teacher_data(teacher_packed):
        teacher_data = unpack_teacher_batch(teacher_packed)
```

File: megatron/core/datasets/blended_megatron_dataset_builder.py [hunk 19]
```diff
@@ -439,9 +439,25 @@ def _build_megatron_dataset_splits(
...
             if split[i] is not None:
-                beg = int(round(split[i][0] * float(num_elements)))
-                end = int(round(split[i][1] * float(num_elements)))
-                split_indices.append(numpy.arange(start=beg, stop=end, step=1, dtype=numpy.int32))
+                beg = int(math.floor(split[i][0] * float(num_elements)))
+                end = int(math.floor(split[i][1] * float(num_elements)))
+
+                beg = max(0, min(beg, num_elements))
+                end = max(beg, min(end, num_elements))
+
```
Comment (himanshu-livup) [inline megatron/core/datasets/blended_megatron_dataset_builder.py]: check if it works without this change with a different split value such as 95,5,0

**Review by himanshu-livup — APPROVED** [review]
